### PR TITLE
Change download file name to match the installer URL

### PR DESCRIPTION
**Pull request by JohnMcPMS**
## Change
Change the installer download file from:
```
%TEMP%\WinGet\<ID>.<Version>.<Installer Extension>
```
to
```
%TEMP%\WinGet\<ID>.<Version>\<Installer URL File Stem>.<Installer Extension>
```

Example:
```
"%TEMP%\WinGet\Notepad++.Notepad++.8.1.9.1.exe"
```
becomes
```
"%TEMP%\WinGet\Notepad++.Notepad++.8.1.9.1\npp.8.1.9.1.Installer.x64.exe"
```

In addition, this change uses:
```
%TEMP%\WinGet\<ID>.<Version>\<Installer SHA256 hash>
```
as the name of the file before hash validation.  This prevents the potential for any user input to create an executable file name since the SHA256 hash must be a hexadecimal string.

## Validation
Existing regression tests and manual runs to ensure consistency for golden path.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/1722)

### CODE REVIEW

File: src/AppInstallerCLICore/Workflows/DownloadFlow.cpp [hunk 1]
```diff
@@ -20,8 +20,22 @@ namespace AppInstaller::CLI::Workflow
...
             const auto& manifest = context.Get<Execution::Data::Manifest>();
+
             std::filesystem::path tempInstallerPath = Runtime::GetPathTo(Runtime::PathName::Temp);
             tempInstallerPath /= Utility::ConvertToUTF16(manifest.Id + '.' + manifest.Version);
+
+            // Get file name from download URI
+            std::filesystem::path filename = GetFileNameFromURI(context.Get<Execution::Data::Installer>()->Url);
+
+            // Assuming that we find a stem value in the URI, use it
+            if (filename.has_stem())
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/DownloadFlow.cpp:31]: >filename [](http://example.com/codeflow?start=16&length=8)

If the filename happens to be ".exe", has_stem() will return true and stem() will just return ".exe", I think that's not what we wanted?

Comment (JohnMcPMS) [reply src/AppInstallerCLICore/Workflows/DownloadFlow.cpp:31]: Are you concerned about the "it already has an executable extension" or the "that is a weird file name" or the "we will create an even weirder file name"?

I don't think we care about weird file names; this is more in line with the name that the developer created (although maybe not perfectly).

As to the file having an executable extension before being hashed, you could already do that by setting `PackageVersion: exe`.  If we want to truly avoid that then we will have to change the code a bit more to use the hash as the temp file name.  But then the "rename with extension" code will have to generate the whole name.  In any case I see it as defense in depth of the general case, not a true security barrier to a dedicated attacker.  But if you feel strongly we can discuss it more.

Comment (yao-msft) [reply src/AppInstallerCLICore/Workflows/DownloadFlow.cpp:31]: I don't feel strongly about the concern. It just does not match what the comment says at the top of the method. So just wanted to confirm we are ok with the less common cases.

**Review by yao-msft — APPROVED** [review]

**Comment by yao-msft** [comment]
Thanks for fixing! Looks good to me.

**Pull request merged**